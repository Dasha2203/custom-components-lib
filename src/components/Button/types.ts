import { HTMLAttributes } from 'react';

export const buttonColor = [
  'primary',
  'secondary',
  'success',
  'error',
] as const;
export type ButtonColor = (typeof buttonColor)[number];

export const buttonVariant = ['text', 'outlined', 'contained'] as const;
export type ButtonVariant = (typeof buttonVariant)[number];

export const buttonSizes = ['small', 'medium', 'large'] as const;
export type ButtonSize = (typeof buttonSizes)[number];

export type ButtonCommonProps = {
  variant?: ButtonVariant;
  size?: ButtonSize;
  color?: ButtonColor;
};

type BtnProps = ButtonCommonProps &
  HTMLAttributes<HTMLButtonElement> & {
    onClick?: React.MouseEventHandler<HTMLButtonElement>;
    disabled?: boolean;
  };

type AnchorProps = ButtonCommonProps &
  HTMLAttributes<HTMLAnchorElement> & {
    href: string;
    target?: '_blank' | '_self' | '_parent' | '_top';
  };

export type ButtonProps = BtnProps | AnchorProps;
