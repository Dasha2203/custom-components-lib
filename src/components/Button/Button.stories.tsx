import { action } from '@storybook/addon-actions';
import type { Meta, StoryObj } from '@storybook/react';
import Button from './Button';
import { ButtonColor, ButtonVariant } from './types';

export const ActionsData = {
  onClick: action('button-click'),
};

const meta = {
  component: Button,
  title: 'Button',
  tags: ['autodocs'],
  excludeStories: /.*Data$/,
  args: {
    ...ActionsData,
  },
} satisfies Meta<typeof Button>;

export default meta;
type Story = StoryObj<typeof meta>;

const DefaultArgs = {
  children: 'Click Me',
};

const createButtonStory = (variant: ButtonVariant, color?: ButtonColor) => {
  return {
    args: {
      ...DefaultArgs,
      variant,
      color,
    },
  };
};

export const Contained: Story = createButtonStory('contained', 'primary');
export const SecondaryContained: Story = createButtonStory(
  'contained',
  'secondary'
);
export const SuccessContained: Story = createButtonStory(
  'contained',
  'success'
);
export const ErrorContained: Story = createButtonStory('contained', 'error');
export const DisabledContained: Story = {
  args: {
    ...createButtonStory('contained', 'error').args,
    disabled: true,
  },
};

export const Outlined: Story = createButtonStory('outlined', 'primary');
export const SecondaryOutlined: Story = createButtonStory(
  'outlined',
  'secondary'
);
export const SuccessOutlined: Story = createButtonStory('outlined', 'success');
export const ErrorOutlined: Story = createButtonStory('outlined', 'error');
export const DisabledOutlined: Story = {
  args: {
    ...createButtonStory('outlined', 'error').args,
    disabled: true,
  },
};

export const Text: Story = createButtonStory('text', 'primary');
export const SecondaryText: Story = createButtonStory('text', 'secondary');
export const SuccessText: Story = createButtonStory('text', 'success');
export const ErrorText: Story = createButtonStory('text', 'error');
export const DisabledText: Story = {
  args: {
    ...createButtonStory('text', 'error').args,
    disabled: true,
  },
};
