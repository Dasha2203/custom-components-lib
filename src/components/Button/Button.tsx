import React from 'react';
import clsx from 'clsx';

import * as classes from './style.module.scss';
import { ButtonProps } from './types';

const Button = ({
  children,
  variant = 'contained',
  size = 'medium',
  ...props
}: ButtonProps) => {
  const className = clsx(
    classes.button,
    classes[`button_${variant}`],
    classes[`button_${size}`],
    props.className
  );

  if ('href' in props) {
    const { href, target, ...restProps } = props;

    return (
      <a className={className} href={href} target={target} {...restProps}>
        {children}
      </a>
    );
  }

  return (
    <button className={className} {...props}>
      {children}
    </button>
  );
};

export default Button;
