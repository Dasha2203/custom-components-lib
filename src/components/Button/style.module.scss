.button {
  --variant-textColor: #1976d2;
  --variant-outlinedColor: #1976d2;
  --variant-outlinedBorder: rgba(25, 118, 210, 0.5);
  --variant-containedColor: #fff;
  --variant-containedBg: #1976d2;

  padding: 6px 16px;

  position: relative;
  display: inline-flex;
  align-items: center;
  justify-content: center;
  box-sizing: border-box;
  min-width: 64px;

  background-color: transparent;
  outline: 0;
  cursor: pointer;
  user-select: none;
  vertical-align: middle;
  text-decoration: none;
  color: inherit;
  font-weight: 500;
  font-size: 0.875rem;
  line-height: 1.75;
  letter-spacing: 0.02857em;
  text-transform: uppercase;
  border: 0;
  border-radius: 4px;
  color: var(--variant-containedColor);
  background-color: var(--variant-containedBg);
  box-shadow:
    0px 3px 1px -2px rgba(0, 0, 0, 0.2),
    0px 2px 2px 0px rgba(0, 0, 0, 0.14),
    0px 1px 5px 0px rgba(0, 0, 0, 0.12);
  transition:
    background-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
    box-shadow 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,
    border-color 250ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;

  &_text {
    color: var(--variant-textColor);
    background-color: var(--variant-textBg);
    box-shadow: none;

    &:disabled {
      color: rgba(0, 0, 0, 0.26);
    }
  }

  &_outline {
    padding: 5px 15px;
    border: 1px solid currentColor;
    box-shadow: none;
    border-color: var(--variant-outlinedBorder, currentColor);
    background-color: var(--variant-outlinedBg);
    color: var(--variant-outlinedColor);

    &:disabled {
      border: 1px solid rgba(0, 0, 0, 0.12);
      color: rgba(0, 0, 0, 0.26);
    }
  }

  &_small {
    padding: 4px 10px;
    font-size: 0.8125rem;
  }

  &_large {
    padding: 8px 22px;
    font-size: 0.9375rem;
  }

  &:disabled {
    box-shadow: none;
    pointer-events: none;
    cursor: default;
  }

  &:hover {
    --variant-containedBg: #1565c0;
    --variant-textBg: rgba(25, 118, 210, 0.04);
    --variant-outlinedBorder: #1976d2;
    --variant-outlinedBg: rgba(25, 118, 210, 0.04);
  }

  &:not(&_outline, &_text) {
    &:active {
      box-shadow:
        0px 5px 5px -3px rgba(0, 0, 0, 0.2),
        0px 8px 10px 1px rgba(0, 0, 0, 0.14),
        0px 3px 14px 2px rgba(0, 0, 0, 0.12);
    }

    &:disabled {
      color: rgba(0, 0, 0, 0.26);
      background-color: rgba(0, 0, 0, 0.12);
    }
  }
}
