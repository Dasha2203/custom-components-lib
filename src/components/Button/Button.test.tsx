import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom';
import Button from './Button';

describe('Test button', () => {
  test('Should render button with text', () => {
    render(<Button>Click me</Button>);
    expect(screen.getByText('Click me')).toBeInTheDocument();
  });

  test('should execute onClick function', () => {
    const handleClick = jest.fn();
    render(<Button onClick={handleClick}>Click</Button>);

    fireEvent.click(screen.getByText('Click'));
    expect(handleClick).toHaveBeenCalledTimes(1);
  });

  test('Should show disabled button', () => {
    render(<Button disabled>Disabled</Button>);
    expect(screen.getByRole('button')).toBeDisabled();
  });

  test('Disabled button shouldn`t execute onclick method', () => {
    const handleClick = jest.fn();
    render(<Button disabled>Click</Button>);
    fireEvent.click(screen.getByText('Click'));

    expect(screen.getByRole('button')).toBeDisabled();
    expect(handleClick).toHaveBeenCalledTimes(0);
  });

  test('Contained Button snapshot', () => {
    const { asFragment } = render(<Button>Click me</Button>);

    expect(asFragment()).toMatchSnapshot();
  });

  test('Outlined Button snapshot', () => {
    const { asFragment } = render(<Button variant="outlined">Click me</Button>);

    expect(asFragment()).toMatchSnapshot();
  });

  test('Text Button snapshot', () => {
    const { asFragment } = render(<Button variant="text">Click me</Button>);

    expect(asFragment()).toMatchSnapshot();
  });

  it('Should contain class for contained button', () => {
    render(<Button variant="contained">Click me</Button>);
    const button = screen.getByText('Click me');

    expect(button).toHaveClass('button_contained');
  });

  it('Should contain class for outlined button', () => {
    render(<Button variant="outlined">Click me</Button>);
    const button = screen.getByText('Click me');

    expect(button).toHaveClass('button_outlined');
    expect(button).toHaveClass('button');
  });

  it('Should contain class for text button', () => {
    render(<Button variant="text">Click me</Button>);
    const button = screen.getByText('Click me');

    expect(button).toHaveClass('button_text');
    expect(button).toHaveClass('button');
  });

  it('should apply additional className prop', () => {
    render(<Button className="extra-class">Click me</Button>);
    const button = screen.getByText('Click me');

    expect(button).toHaveClass('extra-class');
  });

  it('should render an anchor tag when href is provided', () => {
    render(<Button href="https://example.com">Click me</Button>);
    const link = screen.getByRole('link');

    expect(link).toBeInTheDocument();
    expect(link).toHaveAttribute('href', 'https://example.com');
  });

  it('should render button with the correct variant and size classes', () => {
    render(
      <Button variant="outlined" size="small">
        Click me
      </Button>
    );
    const button = screen.getByText('Click me');

    expect(button).toHaveClass('button_outlined');
    expect(button).toHaveClass('button_small');
    expect(button).toHaveClass('button');
  });
});
