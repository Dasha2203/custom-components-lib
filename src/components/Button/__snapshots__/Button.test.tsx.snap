// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test button Contained Button snapshot 1`] = `
<DocumentFragment>
  <button
    class="button button_contained button_medium button_primary"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`Test button Outlined Button snapshot 1`] = `
<DocumentFragment>
  <button
    class="button button_outlined button_medium button_primary"
  >
    Click me
  </button>
</DocumentFragment>
`;

exports[`Test button Text Button snapshot 1`] = `
<DocumentFragment>
  <button
    class="button button_text button_medium button_primary"
  >
    Click me
  </button>
</DocumentFragment>
`;
